src/aapns/connection.py

```diff
@@ -0,0 +1,405 @@
+        finally:
+            self.closing = self.closed = True
+            for sid, ch in self.channels.items():
+                if ch.fut and not ch.fut.done():
+                    ch.fut.set_exception(Closed())
+
+    async def post(self, req: "Request") -> "Response":
+        assert len(req.body) <= MAX_NOTIFICATION_PAYLOAD_SIZE
+
+        now = time()
+        if now > req.deadline:
+            raise Timeout()
+        if self.closing or self.closed:
+            raise Closed()
+        if self.blocked:
+            raise Blocked()
+
+        try:
+            sid = self.conn.get_next_available_stream_id()
+        except h2.NoAvailableStreamIDError:
+            # As of h2-3.2.0 this is permanent.
+            # FIXME ought we mark the connection as closing?
+            raise Closed()
+
+        assert sid not in self.channels
+
+        self.last_new_sid = sid
+
+        ch = self.channels[sid] = Channel(sid, None, [])
+        self.conn.send_headers(
+            sid, req.header_with(self.host, self.port), end_stream=False
+        )
+        # FIXME don't we have to increment global flow control
+        # also on reception of stream X data?
+        self.conn.increment_flow_control_window(2 ** 16, stream_id=sid)
+        self.conn.send_data(sid, req.body, end_stream=True)
+        self.please_write.set()
+
+        try:
+            while not self.closing:
+                ch.fut = Future()
+                with suppress(TimeoutError):
+                    await wait_for(ch.fut, req.deadline - now)
+                now = time()
+                if now > req.deadline:
+                    raise Timeout()
+                for event in ch.ev:
+                    if isinstance(event, h2.events.ResponseReceived):
+                        ch.header = dict(event.headers)
+                    elif isinstance(event, h2.events.DataReceived):
+                        ch.body += event.data
+                    elif isinstance(event, h2.events.StreamEnded):
+                        return Response.new(ch.header, ch.body)
+            else:
+                raise Closed()
+        finally:
+            del self.channels[sid]
+
+    async def background_write(self):
+        try:
+            while not self.closed:
+                data = None
+
+                while not data:
+                    if self.closed:
+                        return
+
+                    if data := self.conn.data_to_send():
+                        self.w.write(data)
+                        last_sid = self.last_new_sid
+                        await self.w.drain()
+                        self.last_sent_sid = last_sid
+                    else:
+                        await self.please_write.wait()
+                        self.please_write.clear()
+
+        except Exception:
+            logger.exception("background write task died")
+        finally:
+            self.closed = True
+
+
+class Blocked(Exception):
```

Comment:
how about moving the exceptions away from this file?

Reply:
Done :)